sklearn/model_selection/tests/test_validate.py

```diff
@@ -0,0 +1,688 @@
+                               n_redundant=0, n_classes=2,
+                               n_clusters_per_class=1, random_state=0)
+    estimator = MockImprovingEstimator(20)
+    assert_raises(ValueError, learning_curve, estimator, X, y, cv=3,
+                  train_sizes=[0, 1])
+    assert_raises(ValueError, learning_curve, estimator, X, y, cv=3,
+                  train_sizes=[0.0, 1.0])
+    assert_raises(ValueError, learning_curve, estimator, X, y, cv=3,
+                  train_sizes=[0.1, 1.1])
+    assert_raises(ValueError, learning_curve, estimator, X, y, cv=3,
+                  train_sizes=[0, 20])
+    assert_raises(ValueError, learning_curve, estimator, X, y, cv=3,
+                  train_sizes=[1, 21])
+
+
+def test_learning_curve_remove_duplicate_sample_sizes():
+    X, y = make_classification(n_samples=3, n_features=1, n_informative=1,
+                               n_redundant=0, n_classes=2,
+                               n_clusters_per_class=1, random_state=0)
+    estimator = MockImprovingEstimator(2)
+    train_sizes, _, _ = assert_warns(
+        RuntimeWarning, learning_curve, estimator, X, y, cv=3,
+        train_sizes=np.linspace(0.33, 1.0, 3))
+    assert_array_equal(train_sizes, [1, 2])
+
+
+def test_learning_curve_with_boolean_indices():
+    X, y = make_classification(n_samples=30, n_features=1, n_informative=1,
+                               n_redundant=0, n_classes=2,
+                               n_clusters_per_class=1, random_state=0)
+    estimator = MockImprovingEstimator(20)
+    cv = KFold(n_folds=3)
+    train_sizes, train_scores, test_scores = learning_curve(
+        estimator, X, y, cv=cv, train_sizes=np.linspace(0.1, 1.0, 10))
+    assert_array_equal(train_sizes, np.linspace(2, 20, 10))
+    assert_array_almost_equal(train_scores.mean(axis=1),
+                              np.linspace(1.9, 1.0, 10))
+    assert_array_almost_equal(test_scores.mean(axis=1),
+                              np.linspace(0.1, 1.0, 10))
+
+
+def test_validation_curve():
+    X, y = make_classification(n_samples=2, n_features=1, n_informative=1,
+                               n_redundant=0, n_classes=2,
+                               n_clusters_per_class=1, random_state=0)
+    param_range = np.linspace(0, 1, 10)
+    with warnings.catch_warnings(record=True) as w:
+        train_scores, test_scores = validation_curve(
+            MockEstimatorWithParameter(), X, y, param_name="param",
+            param_range=param_range, cv=2
+        )
+    if len(w) > 0:
+        raise RuntimeError("Unexpected warning: %r" % w[0].message)
+
+    assert_array_almost_equal(train_scores.mean(axis=1), param_range)
+    assert_array_almost_equal(test_scores.mean(axis=1), 1 - param_range)
+
+
+def test_check_is_permutation():
+    p = np.arange(100)
+    assert_true(_check_is_permutation(p, 100))
+    assert_false(_check_is_permutation(np.delete(p, 23), 100))
+
+    p[0] = 23
+    assert_false(_check_is_permutation(p, 100))
```

Comment:
This isn't the strongest test in the world, it looks like it would perfectly pass for the implementation `lambda x, n: x == np.arange(n)` but it was like this before...